File: Medium/Count digit groupings of a number/count-digit-groupings-of-a-number.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;
class GFG
{
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine().trim());
        while(T-->0)
        {
            String str = br.readLine().trim();
            Solution ob = new Solution();
            int ans = ob.TotalCount(str);
            System.out.println(ans);           
        }
    }
}

// } Driver Code Ends


//User function Template for Java

class Solution {
    
    public int solve(int idx, String s, Integer[][] dp, int sum) {
        if(idx >= s.length()) return 1;
        if(dp[idx][sum] != null) return dp[idx][sum];
        
        int currSum = 0;
        dp[idx][sum] = 0;
        
        for(int i = idx; i<s.length(); i++) {
            currSum += (s.charAt(i) - '0');
            if(sum <= currSum) dp[idx][sum] += solve(i+1, s, dp, currSum);
        }
        
        return dp[idx][sum];
    }
    
    public int TotalCount(String str) {
        Integer[][] dp = new Integer[str.length()][902];
        return solve(0, str, dp, 0);
    }
}

File: Medium/Count digit groupings of a number/count-digit-groupings-of-a-number.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution{
	public:
	int solve(int idx, string s, vector<vector<int>> &dp, int sum) {
	    if(idx >= s.size()) return 1;
	    if(dp[idx][sum] != -1) return dp[idx][sum];
	    
	    int currSum = 0;
	    dp[idx][sum] = 0;
	    
	    for(int i = idx; i<s.size(); i++) {
	        currSum += (s[i] - '0');
	        if(currSum >= sum) dp[idx][sum]  += solve(i+1, s, dp, currSum);
	    }
	    
	    return dp[idx][sum];
	}
	
	int TotalCount(string str) {
	    int n = str.size();
	    vector<vector<int>> dp(n, vector<int>(902, -1));
	    return solve(0, str, dp, 0);
	}

};

//{ Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		string str;
		cin >> str;
		Solution ob;
		int ans = ob.TotalCount(str);
		cout << ans <<"\n";
	}
	return 0;
}
// } Driver Code Ends